docs/2023/puzzles/day14.md

```diff
@@ -2,10 +2,357 @@ import Solver from "../../../../../website/src/components/Solver.js"
+
+To solve Part 1, we do not even need to simulate the movement of the rocks - that is, we do not need to compute the state of the model of the platform after the tilting was performed. We can iterate from the top (north) of the field to the bottom, one line at a time, and use caching to remember where the rocks can fall. The algorithm is as follows:
+
+```scala
+def totalLoading(lines: List[String]): Int =
+  var loading = 0
+  val whereCanIFall = collection.mutable.Map.empty[Int, Int]
+  val totalRows = lines.size
+  for
+    (line, row) <- lines.zipWithIndex
+    (char, col) <- line.zipWithIndex
+  do char match
+    case 'O' =>
+      val fallRow = whereCanIFall.getOrElseUpdate(col, 0)
+      loading += totalRows - fallRow
+      whereCanIFall(col) = fallRow + 1
+    case '#' =>
+      whereCanIFall(col) = row + 1
+    case '.' =>
+  loading
+```
+
+## Part 2 Solution
+Part 2 is much more involved. Here, we are required to tilt the field in all four directions - North, West, South and East - in turn, in many cycles. The objective is to calculate the same loading metric after 1 billion cycles.
+
+The approach from Part 1 doesn't work here: we now need to modify our model state. The reason we didn't have to do it in Part 1 is that our model evolves in only a single step, and we only need one metric from the end state, so we don't have to calculate the entire final state. In Part 2, however, the model evolves in many steps, and each successive step depends on the previous step. So, we need to calculate the entire final state.
+
+### The Model
+Actually, we already know how to do the Northern tilt - similarly to Part 1. We do not want to re-implement the tilts for the other directions - instead, we want to use the Northern tilt as a building block for the other tilts. So, we need to be able to rotate our model in all four directions, and then tilt it North.
+
+We can represent the model as a 2D array of chars:
+
+```scala
+class Model(state: String):
+  private var dataN: Array[Array[Char]] = Array.empty
+  setState(state)
+
+  def setState(newState: String): Unit =
+    val lines = newState.split('\n').toList
+    val extentX = lines(0).length
+    val extentY = lines.length
+    val array = Array.fill(extentX)(Array.fill(extentY)('.'))
+    for
+      (line, y) <- lines.zipWithIndex
+      (char, x) <- line.zipWithIndex
+    do array(x)(y) = char
+    dataN = array
+
+  override def toString =
+    val sb = new StringBuilder
+    for y <- (0 until extentY).toIterator do
+      for x <- (0 until extentX).toIterator do
+        sb.append(this(x, y))
+      sb.append('\n')
+    sb.toString
```

Comment:
```suggestion
    dataN.map(_.mkString).mkString("\n")
```
?

Reply:
`toString` here returns the *rotated* representation of the platform (see `this(x, y)`), as specified by the `rotation` variable`. `dataN.map(_.mkString).mkString("\n")` is not equivalent, as it always returns the default rotation of the platform - facing North.